### PR TITLE
[#10] Introduce a handler for sampling the ~WAIT line and appying wait states.

### PR DESCRIPTION
**Pull request by kosarev**
Based on patches by Simon Owen.

### CODE REVIEW

**Comment by kosarev** [comment]
@simonowen Simon, would this be enough to support SAM Coupé delays?

**Comment by kosarev** [comment]
Updated to call `on_wait()` for i8080.

**Comment by simonowen** [comment]
I've made a small change to separate the MREQ and IORQ cases, since it needs to know whether it's an address or port to apply contention. An address of 00F8 may be uncontended ROM, but it's a contended I/O port. You may prefer to keep a single on_wait handler with a parameter instead.

I've tested with my changes and adjusted the alignment of my I/O contention to match the new CPU changes, and all is working well  :)

File: z80.h [hunk 1]
```diff
@@ -310,14 +310,16 @@ class root {
...
+        self().on_wait_mreq(addr);
         self().on_tick(1); }
     void on_write_cycle_extra_2t() {
         self().on_tick(2); }
 
-    // TODO: Test on_wait() calls.
-    void on_wait(fast_u16 addr) {
+    void on_wait_mreq(fast_u16 addr) {
         unused(addr); }
+    void on_wait_iorq(fast_u16 port) {
```
Comment (kosarev) [inline z80.h:320]: What if `on_rw_wait()` and `on_io_wait()` then? `mreq` and `iorq` refer to specific chip pins, and I'm not sure they are relevant here.

Comment (simonowen) [reply z80.h:320]: The SAM ASIC and Spectrum ULA do monitor CPU lines to detect some events. Page 195 of "The ZX Spectrum ULA" book says "The ULA uses the /IOREQ signal (the result of OR gaming /IORQ with A0) to indicate that the T2 of a ULA port access has begun".

The SAM Coupé schematic shows pins on the ASIC for MREQ and IOREQ, which suggests it's using them directly:
https://velesoft.speccy.cz/samcoupe/sam-schematic/samcoupe-orginal-schematic-mid-bw.png

I'm happy to go with your choice of names if the memory and I/O events can be distinguished by the handlers.

Comment (kosarev) [reply z80.h:320]: Yep, this makes sense if want to go that low level right away. Maybe `on_mreq_wait()` and `on_iorq_wait()` then to follow the existing naming convention and so it doesn't sound like we wait for ~MREQ and ~IORQ?

Comment (simonowen) [reply z80.h:320]: I'm happy to go with `on_mreq_wait()` and `on_iorq_wait()` -- they do make more sense.

I was thinking about potential alternative implementations...  The pure approach might be to have just a single `on_wait()` handler, with a parameter reporting the state of some lines. I've made an assumption only MREQ and IORQ are likely to be of interest, but it's a mechanism that could pass others if needed.

MREQ and IORQ are mutually exclusive so only one of them will be set on every call, and which one is known at the call site. That would give a constant value that the compiler should optimise out from any inline handler check in the derived class. I think it should be as efficient as using separate handlers in non-debug code.  Adding support for additional lines would require some runtime masking before each call, but maybe that's an optimisation to worry about if it happens.

Comment (kosarev) [reply z80.h:320]: Yes, that all sounds familiar as I had to make lots of such decisions previously for this project, and ended up with the most specific handlers to be the most basic ones, so `on_get_b()` and Co. are the primitives and `on_get_reg()` is built on top of them. Similarly, we introduce `on_mreq_wait()` and `on_iorq_wait()` here as the most basic and specific handlers for our immediate needs and postpone make decisions about generic versions, such as something like `on_wait(cycle_type)`, to when we have a real use case for it and therefore some better understanding of what would work best. Thanks to the chosen approach, it's never late to add new handlers without breaking existing code.

**Comment by kosarev** [comment]
Updated to rename and test the new handlers.

**Comment by simonowen** [comment]
Latest changes confirmed as tested and working on my side too.

**Pull request merged**